### ISSUE TITLE
Nesting Futures.transform() eats StackOverflowError and causes the Future to hang forever

### ISSUE THREAD

**Issue by marcinkosiba** [google/guava#2254]
Hey!
I found that the following test will fail with a TimeoutException (the Future hangs indefinitely): 
```java
  static final int REPS = 10000;
  @Test
  public void hangs() throws InterruptedException, TimeoutException {
    SettableFuture<Integer> settableFuture = SettableFuture.create();
    ListenableFuture<Integer> chained = settableFuture;
    for (int i = 0; i < REPS; ++i) {
      chained = Futures.transform(chained, Functions.<Integer>identity());
    }
    settableFuture.set(0);

    Futures.get(chained, 10, TimeUnit.SECONDS, TimeoutException.class);
  }
```
But if I change REPS to be something smaller, like 1000 (this will depend on the size of the stack) then the test passes. I think the root cause is is that the exception can't 'surface' to the final `chained` Future due to a StackOverflowError caused by repeated DirectExecutor.execute application.
 
My logging config may be screwed up, but I didn't see a log for this or anything, which made finding this a bit of a chore. I checked that this happens with Guava 18 and 19. Apologies if this is a KP or already filed (I tried searching Issues, but nothing related pops up).

**Comment by cpovirk** [comment]
Thanks.

This doesn't entirely shock me, but I don't know that we've discussed it before.

The same problem would exist in `CompletableFuture` if not for [some special handling]. If we want to build something similar, we'll probably need to loosen up the spec for `directExecutor`... though I worry that that will have ill effects for other users who justly rely on the current, documented behavior. Perhaps we'd need another executor entirely, one that we encourage over `directExecutor` unless you really need `directExecutor` specifically. Or maybe we really need to build something into `ListenableFuture` itself -- possibly tricky because its API *requires* an executor that we would be ignoring in some cases. (Or we would need it to recognize a special executor -- impossible to guarantee when other `ListenableFuture` implementations already exist.)

A workaround, of course, is to pass in an executor that will actually hand things off to a new thread. This, of course, has costs....

Have you run across this kind of problem in "real" code, or is it something you discovered during other testing?

(I do wonder what's happening to the `StackOverflowError`....)

[some special handling]: http://concurrency.markmail.org/message/24ptlusoytgpzk6n

**Comment by marcinkosiba** [comment]
Hi Chris!

I think it's perfectly reasonable for you to say that there is a limit to the amount of chaining when using a `directExecutor`. The bit I'm worried about and would love to see resolved is the disappearing `StackOverflowError`. I think the reason it goes missing is because the execution goes along these lines:
 1. Start executing a `Future` from the chain
 1. Execution succeeds, mark it as 'done'
 1. Attempt to execute the next `Future` (use `directExecutor`)
 1. Stack overflow
 1. Mark current future as 'failed', wait, nope, it's done, don't need to do anything.

It looks like the exception gets absorbed into a completed future (`ChainingListenableFuture` catches it by catching all `Throwable` and calls `setException` which then ignores the exception because the `Future` is already in the `COMPLETED` state).
The exception could go to the caller (so calling `.set(0)` would throw a `StackOverflowError`) - anywhere is better than nowhere.

We did indeed hit this in ["real" code](https://github.com/facebook/buck/commit/6ee1b3e9a44c9b451411b0e376f3c7eff438fe2e) (`MoreFututres.chainExceptions` is [essentially Futures.transform]( https://github.com/facebook/buck/blob/master/src/com/facebook/buck/util/concurrent/MoreFutures.java#L169)). The code that was building the chain was silly, so the fix was easy. Like I said at the beginning - the fact that the error was entirely suppressed is the real issue.

Cheers!
  Martin

**Comment by cpovirk** [comment]
Thanks -- and sorry for the trouble this caused you.

If we're lucky, we can just do something like changing...

```
setException(t);
```
...to...
```
if (!setException(t)) {
  logger.log(..., t);
}

There's still a possibility of stack overflow, though, especially if the logger code has deep nesting. Another option would be to wrap (if necessary) and rethrow the exception. That should require less stack, and eventually the exception should reach the top-level `directExecutor` exception handler... I think? Hmm, or maybe the *top-most* `directExecutor` handler will catch it, in which case we're back in danger of a stack overflow.

I will think about this some more.

**Comment by marcinkosiba** [comment]
No worries! Personally I'd be fine with you going to the uncaught exception handler directly, but I guess that might be a bit extreme.

I didn't get the last bit:
> `in which case we're back in danger of a stack overflow.` 
From my understanding you'd wrap the `StackOverflowError` in something like `FutureExecutionRuntimeException` (which would be re-thrown/ignored by any intermediate futures). That should unwind the stack all the way to the top (which would most likely be a worker thread) where it would manifest as an runtime exception. How could we get another stack overflow in that case?

**Comment by cpovirk** [comment]
Nice: It turns out that a candidate for `FutureExecutionRuntimeException` already exists: It's the `Error` we have! Here's what I have right now. It's enough to propagate the exception all the way back to whoever called `set()` in the first place:

```
@@ -117,7 +117,36 @@
       setException(e.getCause());
     } catch (Throwable t) {
       // This exception is irrelevant in this thread, but useful for the client.
-      setException(t);
+      if (!setException(t)) {
+        /*
+         * Uh-oh. Most likely cause: set() succeeded in storing the value in the output Future, but
+         * when it went to run the Future's listeners, it got a StackOverflowError because of deep
+         * directExecutor() listener nesting. If user code relies on that listener (e.g., to
+         * complete a downstream, transformed Future), then it may hang. We can't do much to fix
+         * that: Anything we do may overflow the stack again. (And of course stack overflows are bad
+         * news in general. For example, we may have overflowed in the middle of defining a class.
+         * If so, that class will never be loadable in this process.) However, we at least want to
+         * show the error to the user somehow. The simplest way forward is to rethrow it. Because it
+         * is an error, it won't be caught and logged by AbstractFuture.executeListener. Instead, it
+         * can propagates through many layers of AbstractTransformFuture up to the root call to
+         * set().
+         *
+         * https://github.com/google/guava/issues/2254
+         */
+        if (t instanceof Error) {
+          throw (Error) t;
+        }
+        /*
+         * TODO(cpovirk): RuntimeException: Consider rethrowing them, too, though it will buy us
+         * only a couple stack frames before AbstractFuture.executeListener catches and tries to log
+         * it. Or wrap them in an Error (though that's kind of scary) or a special-purpose
+         * RuntimeException that our Future implementations know to pass through.
+         */
+        /*
+         * TODO(cpovirk): checked exceptions: Consider restructuring the code so that the compiler
+         * can see that they're not possible here.
+         */
+      }
     }
   }
```

In other words, this test now passes:

```
  public void testTransform_StackOverflow() throws Exception {
    SettableFuture<Object> root = SettableFuture.create();
    ListenableFuture<Object> output = root;
    for (int i = 0; i < 10000; i++) {
      output = Futures.transform(output, identity());
    }
    try {
      root.set("foo");
      fail();
      // TODO(cpovirk): What terrible problems could this cause for subsequent tests? Can we avoid them by running in a different ClassLoader?
    } catch (StackOverflowError expected) {
    }
  }
  // TODO(cpovirk): replicate to transformAsync and also to catching/catchingAsync
```

I haven't sent this for internal review yet. Any opinions on it?

**Comment by marcinkosiba** [comment]
Looks good to me. I think RuntimeException could also mean that the executor failed or something like that, right? In that case I think the only options to make progress are either logging (+ignore) or re-throwing wrapped in an Error (or `FutureExecutionRuntimeException`).

**Comment by cpovirk** [comment]
A `RuntimeException` from the executor would be caught by `AbstractFuture.executeListener` (if that's what you're asking about). 

In working things like that out, my reviewer and I got confused enough that we decided to restructure things further. The diff is a little long to post here, but basically we eliminated nested flow control entirely :) I think we're pretty happy with it, so once I replicate the code and its tests to `Futures.catching`, you should see the commit. It's always nice when a bug report prompts us to simplify things.

**Comment by cpovirk** [comment]
Sorry for the additional delays. I've submitted the fix (52b5ee6), which will be in Guava 20. Thanks again for the report.
